components/zeekopenbsm/src/openbsmconsumer.cpp

```diff
@@ -0,0 +1,363 @@
+      event.header.remote_port =
+          static_cast<std::int64_t>(ntohs(tok.tt.sockinet_ex32.port));
+      event.header.local_address = "0.0.0.0";
+      event.header.local_port = 0;
+    }
+    if (tok.tt.sockinet_ex32.family == 2) {
+      event.header.family = 2;
+    } else if (tok.tt.sockinet_ex32.family == 26) {
+      event.header.family = 10;
+    } else {
+      event.header.family = 0;
+    }
+    break;
+  }
+  }
+}
+
+Status
+OpenbsmConsumer::populateSocketEvent(Event &event,
+                                     const std::vector<tokenstr_t> &tokens) {
+
+  for (const auto &tok : tokens) {
+    if (tok.id == AUT_SOCKUNIX) {
+      // we filter unix sockets out
+      continue;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+    case AUT_HEADER64:
+    case AUT_HEADER32_EX:
+    case AUT_HEADER64_EX:
+      extractHeader(event, tok);
+      break;
+    case AUT_SUBJECT32:
+    case AUT_SUBJECT64:
+    case AUT_SUBJECT32_EX:
+    case AUT_SUBJECT64_EX:
+      extractSubject(event, tok);
+      break;
+    case AUT_RETURN32:
+    case AUT_RETURN64:
+      extractReturn(event, tok);
+      break;
+    case AUT_SOCKINET32:
+    case AUT_SOCKINET128:
+      extractSocketInet(event, tok);
+      break;
+    }
+  } // for loop
+
+  return Status::success();
+}
+
+Status OpenbsmConsumer::parseMessage() {
+
+  Status status;
+  Event event;
+
+  u_char *buffer = nullptr;
+
+  int reclen = au_read_rec(audit_pipe, &buffer);
+
+  if (reclen <= 0) {
+    return Status::failure("Could not openbsm fetch message");
+  }
+
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+  tokens.reserve(12);
+
+  auto event_id = 0;
+  auto bytesread = 0;
+  while (bytesread < reclen) {
+    if (au_fetch_tok(&tok, buffer + bytesread, reclen - bytesread) == -1) {
+      break;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+      event_id = tok.tt.hdr32.e_type;
+      break;
+    case AUT_HEADER32_EX:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER64:
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+  if (event_filter_list.find(event_id) == event_filter_list.end()) {
+    // Return early for unused event IDs.
+    return Status::success();
+  }
+
+  status = populateSocketEvent(event, tokens);
+
+  if (!status.succeeded()) {
+    d->logger.logMessage(IZeekLogger::Severity::Error, status.message());
+
+  } else {
+    std::lock_guard<std::mutex> lock(d->event_list_mutex);
+    d->event_list.push_back(std::move(event));
+    d->event_list_cv.notify_all();
+  }
+  return status;
+}
+
+Status OpenbsmConsumer::fetchEventsFromPipe() {
+
+  audit_pipe = fopen(AUDIT_PIPE_PATH, "r");
+  if (audit_pipe == nullptr) {
+    throw Status::failure("The auditpipe couldn't be opened.");
+  }
+  event_filter_list.insert(AUE_CONNECT);
+  event_filter_list.insert(AUE_BIND);
```

Comment:
Can we do this in the constructor instead? As written it can race with any reads of this variable.